### PR TITLE
C4 mostafahassaan f8 return most voted synonym

### PR DESCRIPTION
**Pull request by mostafahassaan**
issue #79 
@nourhanZ 
@NourhanMohamed 
@m0hamed 
@SalmaFG 
Please review.

### CODE REVIEW

**Comment by m0hamed** [comment]
Yalla bessala 3alnaby awel reopen. Please use a more elegant way for finding the keyword with the maximum votes

File: arability/app/models/keyword.rb [hunk 1]
```diff
@@ -1,3 +1,26 @@
...
+  
+  # Method gets the synonym of a certain word with the highest
+  # number of votes.
+  # params: 
+  #   word: a Keyword to get the highest voted synonym for
+  # return:
+  #   on succes: the highest voted Synonym of word gets returned. If two synonyms have an 
+  #               equal number of votes, the first synonym entered to the list is returned.
+  #   on failure: if word has no synonyms, nothing is returned
+  def self.highest_voted_synonym(word)
```
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb]: change self. to class << self

File: arability/app/models/synonym.rb [hunk 2]
```diff
@@ -1,4 +1,19 @@
...
   attr_accessible :approved, :name
+  has_many :votes
+
+  # Method returns the the number of votes of a certain synonym.
+  # params:
+  #   syn_id: ID of synonym to get the votes number for.
+  # returns:
+  #   on succes: return an int with number of votes of the synonym with syn_id
+  #   on failure: --
+  def self.getVotes(syn_id)
```
Comment (NourhanMohamed) [inline arability/app/models/synonym.rb]: again the class << self thingy
u have a typo on "succes" in line 10, too

File: arability/app/models/keyword.rb [hunk 3]
```diff
@@ -1,3 +1,26 @@
 class Keyword < ActiveRecord::Base
   attr_accessible :approved, :is_english, :name
+  has_many :synonyms
+
```
Comment (salmafg) [inline arability/app/models/keyword.rb]: Author name missing

File: arability/app/models/synonym.rb [hunk 4]
```diff
@@ -1,4 +1,19 @@
 class Synonym < ActiveRecord::Base
-  belongs_to :keyword
+  belongs_to :keywords
   attr_accessible :approved, :name
+  has_many :votes
+
```
Comment (salmafg) [inline arability/app/models/synonym.rb:6]: Author name missing.

File: arability/app/models/keyword.rb [hunk 5]
```diff
@@ -94,6 +94,23 @@ def get_similar_keywords(search_word, categories = [])
...
+    # Description: Method gets the synonym of a certain word with the highest
+    #               number of votes.
+    # params: 
+    #   word: a Keyword to get the highest voted synonym for
+    # return:
+    #   on succes: the highest voted Synonym of word gets returned. If two 
+    #               synonyms have an equal number of votes, the first synonym 
+    #               entered to the list is returned.
+    #   on failure: if word has no synonyms, nothing is returned
+
```
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb]: Remove this empty line, and adjust lines to be less than 80 characters long
also, there is a typo in line104 na2es s f a5er success

File: arability/app/models/keyword.rb [hunk 6]
```diff
@@ -94,6 +94,23 @@ def get_similar_keywords(search_word, categories = [])
...
+    #               number of votes.
+    # params: 
+    #   word: a Keyword to get the highest voted synonym for
+    # return:
+    #   on succes: the highest voted Synonym of word gets returned. If two 
+    #               synonyms have an equal number of votes, the first synonym 
+    #               entered to the list is returned.
+    #   on failure: if word has no synonyms, nothing is returned
+ 
+    def highest_voted_synonym(word)
```
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb]: word is too vague, choose something more descriptive

File: arability/app/models/keyword.rb [hunk 7]
```diff
@@ -94,6 +94,23 @@ def get_similar_keywords(search_word, categories = [])
...
+    # params: 
+    #   word: a Keyword to get the highest voted synonym for
+    # return:
+    #   on success: the highest voted Synonym of word gets returned. If two 
+    #               synonyms have an equal number of votes, the first synonym 
+    #               entered to the list is returned.
+    #   on failure: if word has no synonyms, nothing is returned
+    def highest_voted_synonym(keyword)
+      max_id = Synonym.where(:keyword_id => keyword.id).joins(:votes)
+      .count(:group => "synonym_id").max
```
Comment (m0hamed) [inline arability/app/models/keyword.rb]: please indent this line

File: arability/app/models/synonym.rb [hunk 8]
```diff
@@ -1,7 +1,8 @@
 class Synonym < ActiveRecord::Base
-  belongs_to :keyword
+  belongs_to :keywords
   attr_accessible :approved, :name
   has_many :votes
+
   validates_format_of :name, :with => /^([\u0621-\u0652 ])+$/, :on => :create, :message => "The synonym is not in the correct form"
```
Comment (NourhanMohamed) [inline arability/app/models/synonym.rb:5]: more than 80 characters

File: arability/app/models/synonym.rb [hunk 9]
```diff
@@ -39,4 +40,5 @@ def recordsynonym(syn, key_id, approved = false)
 
  end
```
Comment (NourhanMohamed) [inline arability/app/models/synonym.rb]: remove extra new lines

File: arability/app/models/keyword.rb [hunk 10]
```diff
@@ -22,14 +12,28 @@ class << self
   # Failure:
   #  returns false if the keyword doesnot exist in the database
   #  or if the approval failed to be saved in the database 
-    def approve_keyword(keyword_id)
-      if Keyword.exists?(id: keyword_id)
-        keyword = Keyword.find(keyword_id)
+  class << self
```
Comment (m0hamed) [inline arability/app/models/keyword.rb]: Mirna's documentation should be below class << self

File: arability/app/models/keyword.rb [hunk 11]
```diff
@@ -22,14 +12,28 @@ class << self
   # Failure:
   #  returns false if the keyword doesnot exist in the database
   #  or if the approval failed to be saved in the database 
-    def approve_keyword(keyword_id)
-      if Keyword.exists?(id: keyword_id)
-        keyword = Keyword.find(keyword_id)
```
Comment (m0hamed) [inline arability/app/models/keyword.rb:27]: also why did you remove mirna's code you messed up the merge resolution please be more mindfull of other peoples code

File: arability/app/models/keyword.rb [hunk 12]
```diff
@@ -22,14 +12,28 @@ class << self
...
+
+  has_many :synonyms
+
+  has_and_belongs_to_many :categories
+  # validates_presence_of :name, 
+  #   :message => "You need to enter a keyword to save"
+  # validates_format_of :name, :with => /^([\u0621-\u0652 ]+|[a-zA-z ]+)$/,
+  #   :message => "The keyword may contain only english or only arabic characters"
+  # validates_uniqueness_of :name,
+  #   :message => "This keyword is already in the database"
```
Comment (m0hamed) [inline arability/app/models/keyword.rb]: validations should be at the top

File: arability/app/models/keyword.rb [hunk 13]
```diff
@@ -130,6 +134,23 @@ def get_similar_keywords(search_word, categories = [])
...
+    # params: 
+    #   word: a Keyword to get the highest voted synonym for
+    # return:
+    #   on success: the highest voted Synonym of word gets returned. If two 
+    #               synonyms have an equal number of votes, the first synonym 
+    #               entered to the list is returned.
+    #   on failure: if word has no synonyms, nothing is returned
+    def highest_voted_synonym(keyword)
+      grouped_synonyms = Synonym.where(:keyword_id => keyword.id).joins(:votes).count(:group => "synonym_id")
+      highest_synonym = grouped_synonyms.max_by{|key, count|count}
```
Comment (m0hamed) [inline arability/app/models/keyword.rb]: insert space before and after the curly braces

File: arability/app/models/synonym.rb [hunk 14]
```diff
@@ -1,5 +1,5 @@
 class Synonym < ActiveRecord::Base
-  belongs_to :keyword
+  belongs_to :keywords
```
Comment (m0hamed) [inline arability/app/models/synonym.rb]: it belongs to a single keyword

File: arability/app/models/synonym.rb [hunk 15]
```diff
@@ -27,6 +27,17 @@ def approve_synonym(synonym_id)
       return false
     end
 
+  end
+end
+
+  has_many :votes
```
Comment (m0hamed) [inline arability/app/models/synonym.rb]: why is this repeated

File: arability/app/models/keyword.rb (hunk 13 above)
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb]: and between the | and the second "count"

Comment (NourhanMohamed) [inline arability/app/models/keyword.rb]: max 80 char per line "line149"

File: arability/app/models/keyword.rb [hunk 16]
```diff
@@ -130,6 +132,23 @@ def get_similar_keywords(search_word, categories = [])
...
+    # Description: Method gets the synonym of a certain word with the highest
+    #               number of votes.
+    # params: 
+    #   word: a Keyword to get the highest voted synonym for
+    # return:
+    #   on success: the highest voted Synonym of word gets returned. If two 
+    #               synonyms have an equal number of votes, the first synonym 
+    #               entered to the list is returned.
+    #   on failure: if word has no synonyms, nothing is returned
+    def highest_voted_synonym(keyword)
```
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb:144]: 80 char per line!

File: arability/app/models/keyword.rb [hunk 17]
```diff
@@ -130,6 +132,23 @@ def get_similar_keywords(search_word, categories = [])
...
+      grouped_synonyms = Synonym.where(:keyword_id => keyword.id).joins(:votes).count(:group => "synonym_id")
+      highest_synonym = grouped_synonyms.max_by{ |key, count|count }
```
Comment (NourhanMohamed) [inline arability/app/models/keyword.rb:147]: 7ot space mabeen el bar wel count ely b3deeha

File: arability/app/models/synonym.rb [hunk 18]
```diff
@@ -18,7 +18,7 @@ class << self
   #  returns false if the synonym doesnot exist in the database
   #  or if the approval failed to be saved in the database 
   
-    def approve_synonym(synonym_id)
+  def approve_synonym(synonym_id)
       if Synonym.exists?(id: synonym_id)
```
Comment (NourhanMohamed) [inline arability/app/models/synonym.rb:22]: if block is indented by 2 tabs, it should be indented by only one

File: arability/app/models/keyword.rb [hunk 19]
```diff
@@ -23,7 +23,7 @@ class << self
   #  returns false if the keyword doesnot exist in the database
   #  or if the approval failed to be saved in the database 
     def approve_keyword(keyword_id)
-      if Keyword.exists?(id: keyword_id)
+      if (Keyword.exists?(id: keyword_id))
```
Comment (m0hamed) [inline arability/app/models/keyword.rb]: no need for brackets

File: arability/app/models/synonym.rb [hunk 20]
```diff
@@ -17,8 +17,8 @@ class << self
   # Failure:
   #  returns false if the synonym doesnot exist in the database
   #  or if the approval failed to be saved in the database 
-  
-    def approve_synonym(synonym_id)
+
+  def approve_synonym(synonym_id)
       if Synonym.exists?(id: synonym_id)
```
Comment (m0hamed) [inline arability/app/models/synonym.rb:22]: this if statement is and remainig block is indented alot

Comment (NourhanMohamed) [inline arability/app/models/synonym.rb:22]: @mostafahassaan wenaby tzabat indentation el block b2a 3ashan n merge w n5las!

Comment (mostafahassaan) [inline arability/app/models/synonym.rb:22]: ya ged3an walahy metzabat 3andi, 3amal 2a-push mesh radi yezbot hena

File: arability/app/models/synonym.rb [hunk 21]
```diff
@@ -48,16 +48,14 @@ def recordsynonym(syn, key_id, approved = false)
...
       return false
     else if Keyword.exists?(id: key_id)
-        synew = Synonym.new
-        synew.name = syn
-        synew.keyword_id = key_id 
-        return synew.save
-      else
-        return false
-      end
+          synew = Synonym.new
```
Comment (m0hamed) [inline arability/app/models/synonym.rb]: you fixed one part and destroyed the other

Comment (m0hamed) [inline arability/app/models/synonym.rb]: zabbat deeh we 5alas
